crates/move-stackless-bytecode/src/spec_purity_analysis.rs

```diff
@@ -72,64 +118,138 @@ impl SpecPurityAnalysis {
+        println!(
+            "Starting analysis for function: {}",
+            func_env.get_full_name_str()
+        );
+
+        let mutable_references = self.find_mutable_reference(&func_env, targets);
 
+        let underlying_func_id = targets.get_fun_by_spec(&func_env.get_qualified_id());
         let code = func_target.get_bytecode();
+        let call_operation = if underlying_func_id.is_some() {
+            self.find_operation_in_function(*underlying_func_id.unwrap(), code)
+        } else {
+            None
+        };
 
-        let call_operation = if underlying_func_id.is_some() { self.find_operation_in_function(*underlying_func_id.unwrap(), code) } else { None };
-        let modif_locations = self.find_modifiable_locations_in_function(code, &func_env, targets, &func_target, &env, &call_operation);
+        // Use process_calls instead of find_network_calls
+        let (network_calls, mut_ref_calls) =
+            self.process_calls(code, targets, &func_target, &env, &call_operation);
 
+        let is_spec = targets.is_function_spec(&func_env.get_qualified_id());
         if is_spec {
-            for loc in modif_locations.iter() {
-                env.diag(
-                    Severity::Error,
-                    loc,
-                    "Consider removing non-pure calls form spec",
-                );
+            println!("Inserting errors");
+            if !network_calls.is_empty() {
+                println!("  Found {} network calls", network_calls.len());
+                for loc in network_calls.iter() {
+                    env.diag(
+                        Severity::Error,
+                        loc,
+                        "Spec function is calling a network module",
+                    );
+                }
+            }
+            if !mutable_references.is_empty() {
+                println!("  Found {} direct mutable references", mutable_references.len());
+                for loc in mutable_references.iter() {
+                    env.diag(
+                        Severity::Error,
+                        loc,
+                        "Spec function has a mutable reference parameter",
+                    );
+                }
+            }
+            if !mut_ref_calls.is_empty() {
+                println!("  Found {} calls with mutable references", mut_ref_calls.len());
+                for loc in mut_ref_calls.iter() {
+                    env.diag(
+                        Severity::Error,
+                        loc,
+                        "Spec function is calling a function that uses mutable references",
+                    );
+                }
             }
         }
 
-        PurityVerificationInfo { is_pure: modif_locations.len() == 0 }
+        let all_mutable_refs = !mutable_references.is_empty() || !mut_ref_calls.is_empty();
+
+        PurityVerificationInfo {
+            is_network_call: !network_calls.is_empty(),
+            is_mutable_reference: all_mutable_refs,
```

Comment:
this should be `!mutable_references.is_empty()`, because it's ok for an immutable function to call mutable functions, as long as it doesn't have mutable parameters.